src/tools/r2rdump/GCInfo.cs

```diff
@@ -0,0 +1,370 @@
+            {
+                // Decode prolog information
+                ValidRangeStart = NativeReader.DecodeVarLengthUnsigned(image, gcInfoTypes.NORM_PROLOG_SIZE_ENCBASE, ref bitOffset) + 1;
+                // satisfy asserts that assume m_GSCookieValidRangeStart != 0 ==> m_GSCookieValidRangeStart < m_GSCookieValidRangeEnd
+                ValidRangeEnd = ValidRangeStart + 1;
+            }
+
+            // Decode the offset to the security object.
+            if ((headerFlags & GcInfoHeaderFlags.GC_INFO_HAS_SECURITY_OBJECT) != 0)
+            {
+                SecurityObjectStackSlot = DenormalizeStackSlot(machine, NativeReader.DecodeVarLengthSigned(image, gcInfoTypes.SECURITY_OBJECT_STACK_SLOT_ENCBASE, ref bitOffset));
+            }
+
+            if ((headerFlags & GcInfoHeaderFlags.GC_INFO_HAS_GS_COOKIE) != 0)
+            {
+                GSCookieStackSlot = DenormalizeStackSlot(machine, NativeReader.DecodeVarLengthSigned(image, gcInfoTypes.GS_COOKIE_STACK_SLOT_ENCBASE, ref bitOffset));
+            }
+
+            if ((headerFlags & GcInfoHeaderFlags.GC_INFO_HAS_PSP_SYM) != 0)
+            {
+                PSPSymStackSlot = DenormalizeStackSlot(machine, NativeReader.DecodeVarLengthSigned(image, gcInfoTypes.PSP_SYM_STACK_SLOT_ENCBASE, ref bitOffset));
+            }
+
+            if ((headerFlags & GcInfoHeaderFlags.GC_INFO_HAS_GENERICS_INST_CONTEXT_MASK) != GcInfoHeaderFlags.GC_INFO_HAS_GENERICS_INST_CONTEXT_NONE)
+            {
+                GenericsInstContextStackSlot = DenormalizeStackSlot(machine, NativeReader.DecodeVarLengthSigned(image, gcInfoTypes.GENERICS_INST_CONTEXT_STACK_SLOT_ENCBASE, ref bitOffset));
+            }
+
+            if ((headerFlags & GcInfoHeaderFlags.GC_INFO_HAS_STACK_BASE_REGISTER) != 0)
+            {
+                if (slimHeader)
+                {
+                    StackBaseRegister = DenormalizeStackBaseRegister(machine, 0);
+                }
+                else
+                {
+                    StackBaseRegister = DenormalizeStackBaseRegister(machine, NativeReader.DecodeVarLengthUnsigned(image, gcInfoTypes.STACK_BASE_REGISTER_ENCBASE, ref bitOffset));
+                }
+            }
+
+            if ((headerFlags & GcInfoHeaderFlags.GC_INFO_HAS_EDIT_AND_CONTINUE_PRESERVED_SLOTS) != 0)
+            {
+                SizeOfEditAndContinuePreservedArea = NativeReader.DecodeVarLengthUnsigned(image, gcInfoTypes.SIZE_OF_EDIT_AND_CONTINUE_PRESERVED_AREA_ENCBASE, ref bitOffset);
+            }
+
+            if (hasReversePInvokeFrame)
+            {
+                ReversePInvokeFrameStackSlot = NativeReader.DecodeVarLengthSigned(image, gcInfoTypes.REVERSE_PINVOKE_FRAME_ENCBASE, ref bitOffset);
+            }
+
+            // FIXED_STACK_PARAMETER_SCRATCH_AREA
```

Comment:
The gcinfoencoder.cpp only writes some of the fields when macros like FIXED_STACK_PARAMETER_SCRATCH_AREA and PARTIALLY_INTERRUPTIBLE_GC_SUPPORTED are defined. How would my program know if they're defined?

Reply:
Your program won't. The ready-to-run compiler and executing runtime environment will need to be built with these macros defined identically. You can see which macros get set in https://github.com/dotnet/coreclr/blob/master/clrdefinitions.cmake and a few other files in the repo root (https://github.com/dotnet/coreclr/blob/master/clrfeatures.cmake for example). Neither of the above two flags are set from the looks of it so you can skip. If we ever enable them, that's a breaking change to ready-to-run images so we'd have to do it in a versionable way (version the gcinfo data structures).